### ISSUE TITLE
Training on new Object

### ISSUE THREAD

**Issue by Ixion46** [NVlabs/Deep_Object_Pose#145]
Hey Guys, thanks for sharing your great work.
I want to test that on my own dataset and training data that I created with NDDS. Which Data do I need to export for the training data? The color images and the instance segmentation? Is it important which value the object has in the image of the class segmentation when I read the image as np.array(000000.cs.png)? Does the class segmentation picture need to be adjusted so that the background is white like in this issues https://github.com/NVlabs/Deep_Object_Pose/issues/137 ? Because for me the class segmentation pictures have black background.
I were not able to find answers in the code for my questions. I hope you can help me
Best Regards Ixion

**Comment by TontonTremblay** [comment]
You only need the rgb images. DOPE does not use the segmentation, sorry for the confusion.

**Comment by Ixion46** [comment]
Thanks for your answer, that helps a lot. I usually translate my models to origin and then axis align them, but with one model I do not get accurate results if i plot the pointcloud of the model with the NDDS gt. 
So my question is now when I use this not axis-aligned model , will DOPE give me the bounding box showed on the image (because these are the values of projected cuboid I guess) or the "best" bounding box along the object? And do I need to train with single objects on the images or can I put like multiple different objects in every scene? 
I hope you can help me again
![model](https://user-images.githubusercontent.com/60901212/96659521-6a2ef480-1347-11eb-81fb-e7879315b8a9.png)

**Comment by TontonTremblay** [comment]
DOPE is agnostic of where the bounding box is, so if you exported this 3d model with NDDS and trained DOPE with that, it should work fine. Just make sure the cuboid dimensions are the ones that have the crooked 3d models within.

**Comment by Ixion46** [comment]
Thanks that helps a lot, but do you also have an answer for the 2nd part of the question?
Will it work if I train with 6 annotated objects in any scene?
![bild](https://user-images.githubusercontent.com/60901212/96661033-72892e80-134b-11eb-99bc-88eb1089f666.png)

**Comment by TontonTremblay** [comment]
I see some some objects with symmetries, e.g., cubes and plate. Dealing with symmetries has not been really explored with DOPE. I had some ideas, but they are not straight forwards to implement and test.

**Comment by Ixion46** [comment]
I have planned to show only 3 sides of the cubes and to rotate the plate only in one axis on the training images, maybe it will work. I will test it out and let you know :).

**Comment by TontonTremblay** [comment]
Let me know how it goes!

**Comment by Ixion46** [comment]
The bottle and cup work great so far! I have training data of 40k images, with all 6 objects on all images.
When I trained the bottle and cup , I had 4000 batches with a batchsize of 20. Now when I train on the green cube, I only have 2000 batches with a batchsize of 20. Why does it differ? I only have one folder with the images

**Comment by Ixion46** [comment]
Somehow the loss of the green cube stays at around 0.002 the whole time, but when I try to detect a cube, it does not detect the cube. Is it because I only show the cube from some specific side or will it simply not work nevermind which training images I will create?

**Comment by TontonTremblay** [comment]
The cube is hard because of symmetries... I would debug the belief map of the inference to see what is going on. I shared the code I use in an other thread.

**Comment by Ixion46** [comment]
![belief_tmp_v1](https://user-images.githubusercontent.com/60901212/96951940-f24efe80-14ed-11eb-99f7-c678d6233c33.png)
![belief_tmp_v2](https://user-images.githubusercontent.com/60901212/96951944-f3802b80-14ed-11eb-90b6-fed4af5a5586.png)
![belief_tmp_v3](https://user-images.githubusercontent.com/60901212/96951947-f5e28580-14ed-11eb-8469-c9ef6c9c70fe.png)
![belief_tmp_v4](https://user-images.githubusercontent.com/60901212/96951952-f8dd7600-14ed-11eb-823d-3255fd4e7e25.png)
![belief_tmp_v5](https://user-images.githubusercontent.com/60901212/96951954-fa0ea300-14ed-11eb-937c-c0a2a3009004.png)
![belief_tmp_v6](https://user-images.githubusercontent.com/60901212/96951959-fbd86680-14ed-11eb-93bb-f9a5cea2852d.png)
![belief_tmp_v7](https://user-images.githubusercontent.com/60901212/96951963-fd099380-14ed-11eb-95c1-561f974c5152.png)
![belief_tmp_v8](https://user-images.githubusercontent.com/60901212/96951966-fe3ac080-14ed-11eb-9542-18636be73c49.png)

**Comment by Ixion46** [comment]
The belief maps does look good for me. Maybe I did something else wrong

**Comment by TontonTremblay** [comment]
This looks good to me, I am not sure what is going on, in the config file can you decrease 
`thresh_map: 0.005`
`thresh_points: 0.05`
to the values above and try again?

**Comment by Ixion46** [comment]
Thanks for your fast replies! Still no detection, the belief maps are close to the others.

**Comment by Ixion46** [comment]
![detection](https://user-images.githubusercontent.com/60901212/96952827-05fb6480-14f0-11eb-821e-90822f58d641.png)
I get a detection if I use thresh_map = 0.001 and thresh_points = 0.01

**Comment by TontonTremblay** [comment]
tweaking is important :P This looks very good, how does it look in the real world. 

The first term is used after smoothing has been applied to the belief map. This is to remove local noise. Then if the threshold you find is too low compared to the neighbors, then it is ignored. If you get that value too low you might get a lot of false positive, be careful when playing around on real data. The second term is about the height peak, e.g., the absolute value, it is somewhat related to the previous one. Anyway, not sure why that object behaves like this.

**Comment by Ixion46** [comment]
I have no opportunity to test it on real data at the moment. Maybe in 2 weeks, then I will definitely report to you. Still do not always get a detection, when I checked on 4-5 other images right now.

**Comment by TontonTremblay** [comment]
So in order to display them I normalize the belief maps, which makes the peak  really punchy. But if you remove that, I am sure the peak are quite soft... 

I am sure you can find images of cubes online to try it out.

**Comment by Ixion46** [comment]
I think I went too far :D
![i_went_too_far](https://user-images.githubusercontent.com/60901212/96955240-6ccf4c80-14f5-11eb-812b-dd38ad165c83.png)

I have 2 questions in my mind. Is it okay for you , if I try to combine your work with Densefusion? Amazingly I have good results on the cubes with Densefusion, but not with the other objects and with your work the other way round.

And have you tested a multiple camera setup on dope?

**Comment by TontonTremblay** [comment]
Can you send me a message to jtremblay@nvidia.com ;)

**Comment by Ixion46** [comment]
After a lucky training epoch a got a loss of 0.0015 for the cube and got a detection now in all tested cases, even without adjusting the tresh_map and thresh_points. :)

**Comment by YeeCY** [comment]
@Ixion46 Could you please explain how you import the red and green boxes into the scene?

**Comment by Ixion46** [comment]
Ofcourse. I use the output of the projected points given by the network and project them into the images.

**Comment by YeeCY** [comment]
> Ofcourse. I use the output of the projected points given by the network and project them into the images.

Oh, sorry for my unclear question. Actually, I want to ask how did you import the red and green boxes into UE4 (or you just create this object in UE4) to create the training dataset?

**Comment by Ixion46** [comment]
I think you mean the orange and yellow boxes? These are the predicted outputs from DOPE and are not in UE4

**Comment by YeeCY** [comment]
> I think you mean the orange and yellow boxes? These are the predicted outputs from DOPE and are not in UE4

Nope. I mean the green and red cubes in the image, which is the input to the algorithm, not the yellow and orange bounding boxes predicted by DOPE. I just wonder how did you create the training dataset.

**Comment by Ixion46** [comment]
Ah now I get it. These are objects in our university lab. I scanned these objects with an artec 3D scanner, edited the models with meshlab and then converted the resulting models to .obj and imported them into UE4.

**Comment by YeeCY** [comment]
> Ah now I get it. These are objects in our university lab. I scanned these objects with an artec 3D scanner, edited the models with meshlab and then converted the resulting models to .obj and imported them into UE4.

Okay, I get it.

**Comment by blaine141** [comment]
How big was the dataset that you used when training this and for how many epochs did you train? I am not having much luck training.
